packages/datastore/src/serveradapter.ts

```diff
@@ -0,0 +1,85 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) 2014-2019, PhosphorJS Contributors
+|
+| Distributed under the terms of the BSD 3-Clause License.
+|
+| The full license is in the file LICENSE, distributed with this software.
+|----------------------------------------------------------------------------*/
+
+import {
+  IDisposable
+} from '@phosphor/disposable';
+
+import {
+  ISignal
+} from '@phosphor/signaling';
+
+import {
+  Datastore
+} from './datastore';
+
+/**
+ * An interface for a bridge between a datastore and a patch server.
+ * Patch servers should expose an interface like this one, which mediates
+ * outgoing patches from a given store, and delivers incoming patches to a
+ * store.
+ */
+export
+interface IServerAdapter extends IDisposable {
+  /**
+   * Broadcast a transaction from a datastore to collaborators.
+   *
+   * @param transaction - the transaction to broadcast to collaborators.
+   *
+   * #### Notes
+   * This is expected to be called by a datastore, and not by any other
+   * user. Direct invocations of this function may have unexpected results.
+   */
+  broadcast(transaction: Datastore.Transaction): void;
+
+  /**
+   * Undo a transaction by id. This sends an undo message to the patch server,
+   * but the undo is not actually done until the datastore recieves the
+   * corresponding transaction and applies it.
+   *
+   * @param id: the transaction to undo.
+   */
+  undo(id: string): Promise<void>;
+
+  /**
+   * Redo a transaction by id.
+   *
+   * @param id: the transaction to redo.
+   */
+  redo(id: string): Promise<void>;
+
+  /**
+   * A signal that is fired when a transaction is received from the server.
+   * Intended to be consumed by a datastore, though other objects may snoop
+   * on the messages.
+   */
+  received: ISignal<this, IServerAdapter.IReceivedArgs>;
```

Comment:
The thought previously was that messages was more appropriate for the 1-to-1 mapping. Do you think that this snooping will be useful for anything?

Reply:
I don't think the snooping would particularly be very useful, I just found the level of indirection with message passing was (a) kind of tough to follow, and (b) less type-safe than this approach. Previously there was really only one message type here (`datastore-transaction`), but with the proliferation, I thought this was cleaner.

Reply:
The other thing that made me a bit unhappy with the message-passing interface was that with that setup, the datastore and the serveradapter needed references to each other, so there was some awkwardness in them setting `this` as a handlers. With this setup, the serveradapter doesn't need a reference to the datastore.

Reply:
It just feels slightly overkill to use a full-fledged signal for this, but I can't think of anything better at the moment.

Reply:
Yeah, I agree that it is a bit overkill, since it is really only intended to be consumed by one entity, but the only alternative I could think of was a one-off callback slot, which seemed a bit of a hack to me.

Reply:
I'm with Vidar on this, I think the signal is overkill. The datastore can assign itself to the server adapter upon construction. I think this is a case where a circular reference is fine.

Reply:
Okay, fair enough. Do you think we should go back to message-passing, or an explicit API? In general, what do we think about this server adapter API (signal excepted)?

Another thing I was considering (and I know that @vidartf) has done some thinking on this: do we include some API for requesting specific transactions, if some happen to get dropped? If so, at what level would that exist?

Reply:
> do we include some API for requesting specific transactions, if some happen to get dropped? If so, at what level would that exist?

I believe this should be the responsibility of the adapter to ensure transmission of all transactions. While the adapter might want to expose some methods to fetch specific transactions, I don't think this needs (or should) be part of the interface definition.

Reply:
I basically agree with this.

Reply:
Okay, I have removed the signal in favor of explicit callbacks functions.